## Week 5 Homework 

In this homework we'll put what we learned about Spark in practice.

For this homework we will be using the FHV 2019-10 data found here. [FHV Data](https://github.com/DataTalksClub/nyc-tlc-data/releases/download/fhv/fhv_tripdata_2019-10.csv.gz)

### Question 1: 

**Install Spark and PySpark** 

- Install Spark
- Run PySpark
- Create a local spark session
- Execute spark.version.

What's the output?

#### Answer

In [1]:
import pyspark
from pyspark.sql import SparkSession

spark = SparkSession.builder\
    .master('local[*]') \
    .appName('test')\
    .getOrCreate()

spark.version

Setting default log level to "WARN".
To adjust logging level use sc.setLogLevel(newLevel). For SparkR, use setLogLevel(newLevel).


24/02/26 02:44:13 WARN NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable


'3.3.2'

### Question 2: 

**FHV October 2019**

Read the October 2019 FHV into a Spark Dataframe with a schema as we did in the lessons.

Repartition the Dataframe to 6 partitions and save it to parquet.

What is the average size of the Parquet (ending with .parquet extension) Files that were created (in MB)? Select the answer which most closely matches.

- 1MB
- 6MB
- 25MB
- 87MB

#### Answer

In [2]:
!wget https://github.com/DataTalksClub/nyc-tlc-data/releases/download/fhv/fhv_tripdata_2019-10.csv.gz

--2024-02-26 02:46:22--  https://github.com/DataTalksClub/nyc-tlc-data/releases/download/fhv/fhv_tripdata_2019-10.csv.gz
Resolving github.com (github.com)... 140.82.114.3
Connecting to github.com (github.com)|140.82.114.3|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://objects.githubusercontent.com/github-production-release-asset-2e65be/513814948/efdfcf82-6d5c-44d1-a138-4e8ea3c3a3b6?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAVCODYLSA53PQK4ZA%2F20240226%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20240226T024623Z&X-Amz-Expires=300&X-Amz-Signature=5fda29bbc28f12d2eca87262aa0d3196bf4484683f2700aa05ec646ff8e0d766&X-Amz-SignedHeaders=host&actor_id=0&key_id=0&repo_id=513814948&response-content-disposition=attachment%3B%20filename%3Dfhv_tripdata_2019-10.csv.gz&response-content-type=application%2Foctet-stream [following]
--2024-02-26 02:46:23--  https://objects.githubusercontent.com/github-production-release-asset-2e65be/513814948/efdfcf82-6d5c-

In [38]:
!zcat fhv_tripdata_2019-10.csv.gz | head -n 11 

dispatching_base_num,pickup_datetime,dropOff_datetime,PUlocationID,DOlocationID,SR_Flag,Affiliated_base_number
B00009,2019-10-01 00:23:00,2019-10-01 00:35:00,264,264,,B00009
B00013,2019-10-01 00:11:29,2019-10-01 00:13:22,264,264,,B00013
B00014,2019-10-01 00:11:43,2019-10-01 00:37:20,264,264,,B00014
B00014,2019-10-01 00:56:29,2019-10-01 00:57:47,264,264,,B00014
B00014,2019-10-01 00:23:09,2019-10-01 00:28:27,264,264,,B00014
B00021         ,2019-10-01 00:00:48,2019-10-01 00:07:12,129,129,,B00021         
B00021         ,2019-10-01 00:47:23,2019-10-01 00:53:25,57,57,,B00021         
B00021         ,2019-10-01 00:10:06,2019-10-01 00:19:50,173,173,,B00021         
B00021         ,2019-10-01 00:51:37,2019-10-01 01:06:14,226,226,,B00021         
B00021         ,2019-10-01 00:28:23,2019-10-01 00:34:33,56,56,,B00021         

gzip: stdout: Broken pipe


In [41]:
from pyspark.sql import types

filename = 'fhv_tripdata_2019-10.csv.gz'

schema = types.StructType([
    types.StructField('dispatching_base_num', types.StringType(), True),
    types.StructField('pickup_datetime', types.TimestampType(), True),
    types.StructField('dropOff_datetime', types.TimestampType(), True),
    types.StructField('PUlocationID', types.IntegerType(), True),
    types.StructField('DOlocationID', types.IntegerType(), True),
    types.StructField('SR_Flag', types.StringType(), True),
    types.StructField('Affiliated_base_number', types.StringType(), True),
])

df = spark.read \
    .option("header", "true") \
    .schema(schema) \
    .csv(filename)

df = df.repartition(6)

df.write.parquet('fhv/2019/10/')

In [46]:
!ls -ltrh fhv/2019/10/*parquet

-rw-r--r-- 1 minasonbol minasonbol 6.4M Feb 26 03:11 fhv/2019/10/part-00002-afc2ec08-16cc-46fc-b3ed-15e0883f4f63-c000.snappy.parquet
-rw-r--r-- 1 minasonbol minasonbol 6.4M Feb 26 03:11 fhv/2019/10/part-00003-afc2ec08-16cc-46fc-b3ed-15e0883f4f63-c000.snappy.parquet
-rw-r--r-- 1 minasonbol minasonbol 6.4M Feb 26 03:11 fhv/2019/10/part-00001-afc2ec08-16cc-46fc-b3ed-15e0883f4f63-c000.snappy.parquet
-rw-r--r-- 1 minasonbol minasonbol 6.4M Feb 26 03:11 fhv/2019/10/part-00000-afc2ec08-16cc-46fc-b3ed-15e0883f4f63-c000.snappy.parquet
-rw-r--r-- 1 minasonbol minasonbol 6.4M Feb 26 03:11 fhv/2019/10/part-00005-afc2ec08-16cc-46fc-b3ed-15e0883f4f63-c000.snappy.parquet
-rw-r--r-- 1 minasonbol minasonbol 6.4M Feb 26 03:11 fhv/2019/10/part-00004-afc2ec08-16cc-46fc-b3ed-15e0883f4f63-c000.snappy.parquet


### Question 3: 

**Count records** 

How many taxi trips were there on the 15th of October?

Consider only trips that started on the 15th of October.

- 108,164
- 12,856
- 452,470
- 62,610

> [!IMPORTANT]
> Be aware of columns order when defining schema

In [48]:
from pyspark.sql import functions as F

# convert pickup_datetime column to a Date type
df = df.withColumn('pickup_date', F.to_date(df.pickup_datetime))

# filter on the 2019-10-15 + count
df.filter(df.pickup_date == "2019-10-15").count()

62610

### Question 4: 

**Longest trip for each day** 

What is the length of the longest trip in the dataset in hours?

- 631,152.50 Hours
- 243.44 Hours
- 7.68 Hours
- 3.32 Hours


#### Answer

In [62]:
timeFmt = "yyyy-MM-dd HH24:mm:ss"

timeDiff = (F.unix_timestamp('dropOff_datetime', format=timeFmt)
            - F.unix_timestamp('pickup_datetime', format=timeFmt))

df.withColumn("duration", timeDiff).agg({"duration": "max"}).collect()[0][0] / 3600

631152.5

### Question 5: 

**User Interface**

Spark’s User Interface which shows the application's dashboard runs on which local port?

- 80
- 443
- 4040
- 8080

#### Answer

8080

### Question 6: 

**Least frequent pickup location zone**

Load the zone lookup data into a temp view in Spark</br>
[Zone Data](https://github.com/DataTalksClub/nyc-tlc-data/releases/download/misc/taxi_zone_lookup.csv)

Using the zone lookup data and the FHV October 2019 data, what is the name of the LEAST frequent pickup location Zone?</br>

- East Chelsea
- Jamaica Bay
- Union Sq
- Crown Heights North